File: frontend/src/api/nutrition.ts
import { api, apiUpload } from "@/api/client";

export interface FoodFields {
  brand?: string | null;
  serving_size?: string | null;
  calories?: number | null;
  protein_g?: number | null;
  carbs_g?: number | null;
  fat_g?: number | null;
  fiber_g?: number | null;
  sugar_g?: number | null;
  sodium_mg?: number | null;
  ingredients?: string | null;
}

export interface FoodDraft extends FoodFields {
  name: string | null;
  image_url: string | null;
  raw: Record<string, unknown>;
}

export interface Food extends FoodFields {
  id: string;
  name: string;
  image_url: string | null;
  created_at: string;
}

export interface MealEntry {
  id: string;
  food_id: string;
  food_name: string;
  meal_type: string;
  servings: number;
  calories: number | null;
  protein_g: number | null;
  logged_at: string;
}

export interface DayNutrition {
  day: string;
  totals: { calories: number; protein_g: number; carbs_g: number; fat_g: number };
  entries: MealEntry[];
}

export function analyzeLabel(file: File): Promise<FoodDraft> {
  const form = new FormData();
  form.append("file", file);
  return apiUpload<FoodDraft>("/nutrition/analyze", form);
}

export const createFood = (payload: FoodFields & { name: string; image_url?: string | null; raw?: unknown }) =>
  api<Food>("/nutrition/foods", { method: "POST", body: JSON.stringify(payload) });

export const listFoods = () => api<Food[]>("/nutrition/foods");

export const deleteFood = (id: string) =>
  api<void>(`/nutrition/foods/${id}`, { method: "DELETE" });

export const logMeal = (food_id: string, meal_type: string, servings: number, logged_on?: string) =>
  api<MealEntry>("/nutrition/log", {
    method: "POST",
    body: JSON.stringify({ food_id, meal_type, servings, logged_on: logged_on ?? null }),
  });

export const removeLog = (logId: string) =>
  api<void>(`/nutrition/log/${logId}`, { method: "DELETE" });

/** Nutrition for a given day (YYYY-MM-DD); omit for today. */
export const getDay = (on?: string) =>
  api<DayNutrition>(`/nutrition/day${on ? `?on=${on}` : ""}`);

File: backend/app/schemas/nutrition.py
"""Nutrition (foods + meal log) schemas (docs/05, docs/10)."""
from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field


class FoodFields(BaseModel):
    """Editable nutrition fields shared by the OCR draft and the saved food."""

    brand: str | None = None
    serving_size: str | None = None
    calories: float | None = None
    protein_g: float | None = None
    carbs_g: float | None = None
    fat_g: float | None = None
    fiber_g: float | None = None
    sugar_g: float | None = None
    sodium_mg: float | None = None
    ingredients: str | None = None


class FoodDraft(FoodFields):
    """What the OCR endpoint returns for the user to review/name before saving."""

    name: str | None = None
    image_url: str | None = None
    raw: dict = {}


class FoodCreate(FoodFields):
    name: str = Field(min_length=1, max_length=120)
    image_url: str | None = None
    raw: dict = {}


class FoodOut(FoodFields):
    id: uuid.UUID
    name: str
    image_url: str | None
    created_at: datetime


class MealLogCreate(BaseModel):
    food_id: uuid.UUID
    meal_type: str = "snack"  # breakfast|lunch|dinner|snack
    servings: float = Field(default=1, gt=0, le=50)
    logged_on: date | None = None  # defaults to today; set to backfill a past day


class MealLogOut(BaseModel):
    id: uuid.UUID
    food_id: uuid.UUID
    food_name: str
    meal_type: str
    servings: float
    calories: float | None
    protein_g: float | None
    logged_at: datetime


class DayTotals(BaseModel):
    calories: float
    protein_g: float
    carbs_g: float
    fat_g: float


class DayNutrition(BaseModel):
    day: date
    totals: DayTotals
    entries: list[MealLogOut]
